README
pixi-tools
==========

pixi-tools: a set of software to interface with the Raspberry Pi and
PiXi-200 hardware
Copyright (C) 2013 Simon Cantrill

pixi-tools is free software; you can redistribute it and/or modify it under
the terms of the GNU Lesser General Public License as published by the Free
Software Foundation; either version 2.1 of the License, or (at your option)
any later version.

The following components are included:

libpixi/
    A low level c-library.  The library is written in C99, but the headers
    are C89 compatible (except log.h, which is gnu89 compatible).

pio/
    A C program command-line interface that provides hardware access via
    libpixi.

swig/
    Generated python bindings for libpixi. The following modules are available:
    pixitools.pi
        a wrapper of the generic and Raspberry Pi related parts of libpixi
    pixitools.pixi
        a wrapper of the PiXi-200 related parts of libpixi

python/lib/
    Some additional pure python modules.

python/scripts/
    A small selection of scripts that make use of and demonstrate the
    pixitools python package:
    pixi-server:
        Provides access to Pi and PiXi hardware via a web-browser

Python versions 2.7 and >=3.2 are supported.

libpixi and pio have selectable log levels via environment variables.  Use
LIBPIXI_LOG_LEVEL=? and/or PIO_LOG_LEVEL=? where ? is one of all, trace,
debug, info, warn, error, fatal, off.  Example:

root@raspberrypi:~# pio fpga-version
210349040313
root@raspberrypi:~# LIBPIXI_LOG_LEVEL=trace pio fpga-version
pio: trace: libpixi initialised
pio: trace: Found hardware [BCM2708]
pio: trace: Found board revision [0002]
pio: trace: Setting board version to 1
pio: trace: 3 = pixi_open (filename="/dev/spidev0.0", flags=2, mode=0)
pio: debug: Opened SPI name=/dev/spidev0.0 fd=3
pio: trace: pixi_spiReadWrite of fd=3, bufferSize=4
pio: trace: pixi_spiReadWrite of fd=3, bufferSize=4
pio: trace: pixi_spiReadWrite of fd=3, bufferSize=4
pio: debug: Got PiXi FPGA version 2103,4904,0313
pio: debug: Closing SPI device fd=3
210349040313



debian/changelog
pixi-tools (0.6.1-1) unstable; urgency=medium

  * Fix uninitialised struct spi_ioc_transfer in multiRegisterOp
    - The pad value was uninitialised, which may break things on later kernels.

 -- Simon Cantrill <simonjcantrill@gmail.com>  Tue, 08 Sep 2015 18:54:57 +0100

pixi-tools (0.6-1) unstable; urgency=medium

  * More i2c functions

  * More MPU functions, with fix for temperature representation

  * New pio commands:
    - i2c-transfer
    - mpu-read-temp
    - mpu-read-acc
    - mpu-read-acc-x
    - mpu-read-acc-y
    - mpu-read-acc-z
    - mpu-read-gyro
    - mpu-read-gyro-x
    - mpu-read-gyro-y
    - mpu-read-gyro-z
    - mpu-monitor-mag
    - mpu-read-mag
    - mpu-read-mag-x
    - mpu-read-mag-y
    - mpu-read-mag-z
    - mpu-get-acc-scale
    - mpu-set-acc-scale
    - mpu-get-gyro-scale
    - mpu-set-gyro-scale

 -- Simon Cantrill <simonjcantrill@gmail.com>  Sat, 15 Nov 2014 14:44:16 +0000

pixi-tools (0.5.1) unstable; urgency=medium

  * Fix the init script to load FPGA when necessary

  * Add modprobe of i2c-dev and i2c-bcm2708 to init script

  * Don't gzip the example .c code

 -- Simon Cantrill <simonjcantrill@gmail.com>  Tue, 23 Sep 2014 19:31:50 +0100

pixi-tools (0.5) unstable; urgency=medium

  * Many API additions and changes
    - Some APIs (mostly GPIO related) have been renamed
    - From this point, the aim is to retain API and ABI compatibility

  * The default FPGA file is now /etc/pixi-tools/pixi.bin
    - Nothing is installed there by default - the user should make a symlink

  * Flash support has been added

  * Some basic i2c support has been added

  * Some basic MPU support has been added

  * New pio commands:
    - adc-monitor
    - dac-write
    - flash-rdp-res
    - flash-read-id
    - flash-read-status
    - flash-read
    - flash-write
    - flash-write-no-erase
    - flash-erase-sectors
    - flash-erase
    - fpga-dna
    - i2c-read
    - i2c-write
    - monitor-pi-gpio
    - mpu-monitor-temp
    - mpu-monitor-motion
    - spi-transfer

  * SPI channel arguments have been removed from various pio SPI commands

  * Improvements to logging
    - Add macros in libpixi to provide application specific logging
    - Application logging inherits log-level from libpixi (if not specified)

  * pio command handling moved into libpixi
    - Now available for other programs to use

  * Add a pkg-config file
    - /usr/lib/pkgconfig/libpixi.pc

  * Add pixi-cc and pixi-c++ scripts
    - These wrap the gcc and g++ commands with the arguments required for
      using libpixi

  * Some pio extensions are now installed as example code in doc/pixi-tools

  * Improvements to libpixi version handling

  * Improvements to gpio functions
    - No longer implicitly use the wiringPi pin mapping

 -- Simon Cantrill <simonjcantrill@gmail.com>  Sun, 21 Sep 2014 14:38:29 +0100

pixi-tools (0.4preview2-1) unstable; urgency=medium

  * Fixes for reading ADC, setting LCD position

  * Signed packages

 -- Simon Cantrill <simonjcantrill@gmail.com>  Tue, 08 Jul 2014 20:05:54 +0100

pixi-tools (0.4preview1-1) unstable; urgency=medium

  * Snapshot of current development

  * Add log file support

  * Add pio spi-scan command

  * Add pio --help-all option

  * Support PiXi version 2 ADC

  * API improvements:
    - Remove the SpiDevice* parameter from most APIs
    - Remove struct LcdDevice

  * Python improvements:
    - Generate python wrappers that throw exceptions on error
    - Add generated python doc to functions

  * Add some (not yet functional) UART code

  * Add dalek and telescope-remote extensions

 -- Simon Cantrill <simonjcantrill@gmail.com>  Sun, 06 Jul 2014 20:01:52 +0100

pixi-tools (0.3-1) unstable; urgency=medium

  * Configurable UI, with additional capabilities

  * Support for /etc/pixi-init startup hook

 -- Simon Cantrill <simonjcantrill@gmail.com>  Mon, 17 Feb 2014 06:44:59 +0000

pixi-tools (0.2.1-1) unstable; urgency=low

  * Improvements to existing APIs

  * Add libpixi/pixi/simple.h API

  * Improvements to server pages

  * Add rover.c as an example extension

  * Add pixisim library as a crude simulator

 -- Simon Cantrill <simonjcantrill@gmail.com>  Sun, 08 Dec 2013 16:33:23 +0000

pixi-tools (0.2-1) unstable; urgency=low

  * Add real debian packaging infrastructure
    - Install into /usr instead of /usr/local.
    - Use dh_pysupport.

  * Many code changes:
    - Make libpixi a shared library.
    - Add command 'fpga-build-time'.
    - Add command 'adc-read'.
    - Fixes to the dynamically generated HTML.
    - Addition of the 'Sequencer' UI.

 -- Simon Cantrill <simonjcantrill@gmail.com>  Tue, 19 Nov 2013 20:38:32 +0000

pixi-tools (0.1-1) unstable; urgency=low

  * Initial release.

 -- Simon Cantrill <simonjcantrill@gmail.com>  Wed, 18 Sep 2013 19:50:02 +0100


pio/mpu.c
/*
    pixi-tools: a set of software to interface with the Raspberry Pi
    and PiXi-200 hardware
    Copyright (C) 2014 Simon Cantrill

    pixi-tools is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation; either
    version 2.1 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this library; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include <libpixi/pixi/mpu.h>
#include <libpixi/util/file.h>
#include <libpixi/util/io.h>
#include <libpixi/util/string.h>
#include <stdio.h>
#include "common.h"
#include "log.h"


const int interval = 100 * 1000;

static int mpuInit (void)
{
	// FIXME: extend this, and move to libpixi
	return pixi_mpuWriteRegister (MpuPowerManagement1, 0);
}

static int mpuOpenInit (void)
{
	int result = pixi_mpuOpen();
	if (result < 0)
		return result;
	return mpuInit();
}

static int mpuMagOpenInit (void)
{
	int result = mpuOpenInit();
	if (result < 0)
		return result;
	result = pixi_mpuMagOpen();
	if (result < 0)
		pixi_mpuClose();
	return result;
}


static int mpuGetAccelScale (void)
{
	int result = mpuOpenInit();
	if (result < 0)
		return result;

	result = pixi_mpuGetAccelScale();
	if (result >= 0)
		printf ("%d\n", result);

	pixi_mpuClose();

	return result;
}
static int mpuGetAccelScaleFn (const Command* command, uint argc, char* argv[])
{
	LIBPIXI_UNUSED(argv);
	if (argc != 1)
		return commandUsageError (command);

	return mpuGetAccelScale();
}
static Command mpuGetAccelScaleCmd =
{
	.name        = "mpu-get-acc-scale",
	.description = "Get the accelerometer range (g)",
	.usage       = "usage: %s",
	.function    = mpuGetAccelScaleFn
};


static int mpuSetAccelScale (uint scale)
{
	int result = mpuOpenInit();
	if (result < 0)
		return result;

	result = pixi_mpuSetAccelScale (scale);

	pixi_mpuClose();

	return result;
}
static int mpuSetAccelScaleFn (const Command* command, uint argc, char* argv[])
{
	LIBPIXI_UNUSED(argv);
	if (argc != 2)
		return commandUsageError (command);

	uint scale = pixi_parseLong (argv[1]);

	return mpuSetAccelScale (scale);
}
static Command mpuSetAccelScaleCmd =
{
	.name        = "mpu-set-acc-scale",
	.description = "Set the accelerometer range (g)",
	.usage       = "usage: %s 2|4|8|16",
	.function    = mpuSetAccelScaleFn
};


static int mpuGetGyroScale (void)
{
	int result = mpuOpenInit();
	if (result < 0)
		return result;

	result = pixi_mpuGetGyroScale();
	if (result >= 0)
		printf ("%d\n", result);

	pixi_mpuClose();

	return result;
}
static int mpuGetGyroScaleFn (const Command* command, uint argc, char* argv[])
{
	LIBPIXI_UNUSED(argv);
	if (argc != 1)
		return commandUsageError (command);

	return mpuGetGyroScale();
}
static Command mpuGetGyroScaleCmd =
{
	.name        = "mpu-get-gyro-scale",
	.description = "Get the gyroscope range (degrees/sec)",
	.usage       = "usage: %s",
	.function    = mpuGetGyroScaleFn
};


static int mpuSetGyroScale (uint scale)
{
	int result = mpuOpenInit();
	if (result < 0)
		return result;

	result = pixi_mpuSetGyroScale (scale);

	pixi_mpuClose();

	return result;
}
static int mpuSetGyroScaleFn (const Command* command, uint argc, char* argv[])
{
	LIBPIXI_UNUSED(argv);
	if (argc != 2)
		return commandUsageError (command);

	uint scale = pixi_parseLong (argv[1]);

	return mpuSetGyroScale (scale);
}
static Command mpuSetGyroScaleCmd =
{
	.name        = "mpu-set-gyro-scale",
	.description = "Set the gyroscope range (degrees/sec)",
	.usage       = "usage: %s 250|500|1000|2000",
	.function    = mpuSetGyroScaleFn
};


static int mpuMonitorTemp (void)
{
	int result = mpuOpenInit();
	if (result < 0)
		return result;

	result = 0;
	while (true)
	{
		result = pixi_mpuReadRegister16 (MpuTemperatureHigh);
		if (result < 0)
			break;
		int16 raw = result;
		double temp = mpuTemperatureToDegrees (raw);
		printf ("\rhex=0x%04x dec=%6d temperature=%7.3f degrees C", (uint16) raw, raw, temp);
		fflush (stdout);

		usleep (interval);
	}
	printf ("\n");

	pixi_mpuClose();

	return result;
}


static int mpuMonitorTempFn (const Command* command, uint argc, char* argv[])
{
	LIBPIXI_UNUSED(argv);
	if (argc != 1)
		return commandUsageError (command);

	return mpuMonitorTemp();
}
static Command mpuMonitorTempCmd =
{
	.name        = "mpu-monitor-temp",
	.description = "Monitor MPU temperature",
	.usage       = "usage: %s",
	.function    = mpuMonitorTempFn
};


static int mpuReadTemp (void)
{
	int result = mpuOpenInit();
	if (result < 0)
		return result;

	result = pixi_mpuReadRegister16 (MpuTemperatureHigh);
	if (result < 0)
		return result;

	int16 raw = result;
	double temp = mpuTemperatureToDegrees (raw);
	printf ("%.3f\n", temp);

	pixi_mpuClose();

	return result;
}

static int mpuReadTempFn (const Command* command, uint argc, char* argv[])
{
	LIBPIXI_UNUSED(argv);
	if (argc != 1)
		return commandUsageError (command);

	return mpuReadTemp();
}
static Command mpuReadTempCmd =
{
	.name        = "mpu-read-temp",
	.description = "Read MPU temperature (degrees C)",
	.usage       = "usage: %s",
	.function    = mpuReadTempFn
};


static int mpuMonitorMotion (void)
{
	int result = mpuOpenInit();
	if (result < 0)
		return result;

	result = pixi_mpuGetAccelScale();
	if (result < 0)
	{
		pixi_mpuClose();
		return result;
	}
	const double accelScale = result / 32768.0;

	result = pixi_mpuGetGyroScale();
	if (result < 0)
	{
		pixi_mpuClose();
		return result;
	}
	const double gyroScale = result / 32768.0;

	MpuMotion motion;
	result = 0;
	while (true)
	{
		result = pixi_mpuReadMotion (&motion);
		if (result < 0)
			break;
		printf ("\r[%8.4f %8.4f %8.4f]g [%8.3f %8.3f %8.3f]dps",
			accelScale * motion.accel.x, accelScale * motion.accel.y, accelScale * motion.accel.z,
			gyroScale * motion.gyro.x, gyroScale * motion.gyro.y, gyroScale * motion.gyro.z
		);
		fflush (stdout);

		usleep (interval);
	}
	printf ("\n");

	pixi_mpuClose();

	return result;
}


static int mpuMonitorMotionFn (const Command* command, uint argc, char* argv[])
{
	LIBPIXI_UNUSED(argv);
	if (argc != 1)
		return commandUsageError (command);

	return mpuMonitorMotion();
}
static Command mpuMonitorMotionCmd =
{
	.name        = "mpu-monitor-motion",
	.description = "Monitor MPU motion",
	.usage       = "usage: %s",
	.function    = mpuMonitorMotionFn
};


static int mpuMonitorMag (void)
{
	int result = mpuMagOpenInit();
	if (result < 0)
		return result;

	result = 0;
	MpuAxes adjust = {0,0,0};
	result = pixi_mpuReadMagAdjust (&adjust);
	if (result < 0)
		APP_ERROR(-result, "Failed to read magnetometer adjustment values");

	double x = mpuMagGetScale (adjust.x);
	double y = mpuMagGetScale (adjust.y);
	double z = mpuMagGetScale (adjust.z);

	const char* format = "\r[%8.3f %8.3f %8.3f]uT";
	if (pixi_isLocaleEncodingUtf8())
		format = "\r[%8.3f %8.3f %8.3f]μT";

	while (true)
	{
		MpuAxes axes;
		result = pixi_mpuReadMag (&axes);
		if (result < 0)
			break;
		printf (format, x * axes.x, y * axes.y, z * axes.z);
		fflush (stdout);

		usleep (interval);
	}
	printf ("\n");

	pixi_mpuClose();

	return result;
}


static int mpuMonitorMagFn (const Command* command, uint argc, char* argv[])
{
	LIBPIXI_UNUSED(argv);
	if (argc != 1)
		return commandUsageError (command);

	return mpuMonitorMag();
}
static Command mpuMonitorMagCmd =
{
	.name        = "mpu-monitor-mag",
	.description = "Monitor MPU magnetometer",
	.usage       = "usage: %s",
	.function    = mpuMonitorMagFn
};


static int mpuReadAcc (void)
{
	int result = mpuOpenInit();
	if (result < 0)
		return result;

	result = pixi_mpuGetAccelScale();
	if (result < 0)
	{
		pixi_mpuClose();
		return result;
	}
	const double scale = result / 32768.0;

	MpuAxes ax;
	result = pixi_mpuReadAccel (&ax);
	if (result >= 0)
		printf ("%.4f %.4f %.4f\n", scale * ax.x, scale * ax.y, scale * ax.z);

	pixi_mpuClose();

	return result;
}


static int mpuReadAccFn (const Command* command, uint argc, char* argv[])
{
	LIBPIXI_UNUSED(argv);
	if (argc != 1)
		return commandUsageError (command);

	return mpuReadAcc();
}
static Command mpuReadAccCmd =
{
	.name        = "mpu-read-acc",
	.description = "Read accelerometer (g)",
	.usage       = "usage: %s",
	.function    = mpuReadAccFn
};


static int mpuReadAccX (uint address, const Command* command, uint argc, char* argv[])
{
	LIBPIXI_UNUSED(argv);
	if (argc != 1)
		return commandUsageError (command);

	int result = mpuOpenInit();
	if (result < 0)
		return result;

	result = pixi_mpuGetAccelScale();
	if (result < 0)
	{
		pixi_mpuClose();
		return result;
	}
	const double scale = result / 32768.0;

	result = pixi_mpuReadRegister16 (address);
	if (result >= 0)
		printf ("%.4f\n", scale * (int16) result);

	pixi_mpuClose();

	return result;
}


static int mpuReadAccXFn (const Command* command, uint argc, char* argv[])
{
	return mpuReadAccX (MpuAccelXHigh, command, argc, argv);
}
static Command mpuReadAccXCmd =
{
	.name        = "mpu-read-acc-x",
	.description = "Read accelerometer x value (g)",
	.usage       = "usage: %s",
	.function    = mpuReadAccXFn
};


static int mpuReadAccYFn (const Command* command, uint argc, char* argv[])
{
	return mpuReadAccX (MpuAccelYHigh, command, argc, argv);
}
static Command mpuReadAccYCmd =
{
	.name        = "mpu-read-acc-y",
	.description = "Read accelerometer y value (g)",
	.usage       = "usage: %s",
	.function    = mpuReadAccYFn
};


static int mpuReadAccZFn (const Command* command, uint argc, char* argv[])
{
	return mpuReadAccX (MpuAccelZHigh, command, argc, argv);
}
static Command mpuReadAccZCmd =
{
	.name        = "mpu-read-acc-z",
	.description = "Read accelerometer z value (g)",
	.usage       = "usage: %s",
	.function    = mpuReadAccZFn
};


static int mpuReadGyro (void)
{
	int result = mpuOpenInit();
	if (result < 0)
		return result;

	result = pixi_mpuGetGyroScale();
	if (result < 0)
	{
		pixi_mpuClose();
		return result;
	}
	const double scale = result / 32768.0;

	MpuAxes ax;
	result = pixi_mpuReadGyro (&ax);
	if (result >= 0)
		printf ("%.3f %.3f %.3f\n", scale * ax.x, scale * ax.y, scale * ax.z);

	pixi_mpuClose();

	return result;
}


static int mpuReadGyroFn (const Command* command, uint argc, char* argv[])
{
	LIBPIXI_UNUSED(argv);
	if (argc != 1)
		return commandUsageError (command);

	return mpuReadGyro();
}
static Command mpuReadGyroCmd =
{
	.name        = "mpu-read-gyro",
	.description = "Read gyroscope (dps)",
	.usage       = "usage: %s",
	.function    = mpuReadGyroFn
};


static int mpuReadGyroX (uint address, const Command* command, uint argc, char* argv[])
{
	LIBPIXI_UNUSED(argv);
	if (argc != 1)
		return commandUsageError (command);

	int result = mpuOpenInit();
	if (result < 0)
		return result;

	result = pixi_mpuGetGyroScale();
	if (result < 0)
	{
		pixi_mpuClose();
		return result;
	}
	const double scale = result / 32768.0;

	result = pixi_mpuReadRegister16 (address);
	if (result >= 0)
		printf ("%.3f\n", scale * (int16) result);

	pixi_mpuClose();

	return result;
}


static int mpuReadGyroXFn (const Command* command, uint argc, char* argv[])
{
	return mpuReadGyroX (MpuGyroXHigh, command, argc, argv);
}
static Command mpuReadGyroXCmd =
{
	.name        = "mpu-read-gyro-x",
	.description = "Read gyroscope x value (dps)",
	.usage       = "usage: %s",
	.function    = mpuReadGyroXFn
};


static int mpuReadGyroYFn (const Command* command, uint argc, char* argv[])
{
	return mpuReadGyroX (MpuGyroYHigh, command, argc, argv);
}
static Command mpuReadGyroYCmd =
{
	.name        = "mpu-read-gyro-y",
	.description = "Read gyroscope y value (dps)",
	.usage       = "usage: %s",
	.function    = mpuReadGyroYFn
};


static int mpuReadGyroZFn (const Command* command, uint argc, char* argv[])
{
	return mpuReadGyroX (MpuGyroZHigh, command, argc, argv);
}
static Command mpuReadGyroZCmd =
{
	.name        = "mpu-read-gyro-z",
	.description = "Read gyroscope z value (dps)",
	.usage       = "usage: %s",
	.function    = mpuReadGyroZFn
};


static int mpuReadMagAdjusted (double values[])
{
	int result = mpuMagOpenInit();
	if (result < 0)
		return result;

	result = 0;
	MpuAxes adjust = {0,0,0};
	result = pixi_mpuReadMagAdjust (&adjust);
	if (result < 0)
		APP_ERROR(-result, "Failed to read magnetometer adjustment values");

	MpuAxes axes;
	result = pixi_mpuReadMag (&axes);
	if (result >= 0)
	{
		values[0] = mpuMagAdjust (adjust.x, axes.x);
		values[1] = mpuMagAdjust (adjust.y, axes.y);
		values[2] = mpuMagAdjust (adjust.z, axes.z);
	}
	else
		APP_ERROR(-result, "Failed to read magnetometer values");

	pixi_mpuClose();

	return result;
}

static int mpuReadMagFn (const Command* command, uint argc, char* argv[])
{
	LIBPIXI_UNUSED(argv);
	if (argc != 1)
		return commandUsageError (command);

	double axes[3];
	int result = mpuReadMagAdjusted (axes);
	if (result < 0)
		return result;

	printf ("%.3f %.3f %.3f\n", axes[0], axes[1], axes[2]);

	return result;
}
static Command mpuReadMagCmd =
{
	.name        = "mpu-read-mag",
	.description = "Read magnetometer",
	.usage       = "usage: %s",
	.function    = mpuReadMagFn
};

static int mpuReadMagAxis (uint axis, const Command* command, uint argc, char* argv[])
{
	LIBPIXI_UNUSED(argv);
	if (argc != 1)
		return commandUsageError (command);

	double axes[3];
	int result = mpuReadMagAdjusted (axes);
	if (result < 0)
		return result;

	printf ("%.3f\n", axes[axis]);

	return result;
}


static int mpuReadMagXFn (const Command* command, uint argc, char* argv[])
{
	return mpuReadMagAxis (0, command, argc, argv);
}
static Command mpuReadMagXCmd =
{
	.name        = "mpu-read-mag-x",
	.description = "Read magnetometer x value",
	.usage       = "usage: %s",
	.function    = mpuReadMagXFn
};


static int mpuReadMagYFn (const Command* command, uint argc, char* argv[])
{
	return mpuReadMagAxis (1, command, argc, argv);
}
static Command mpuReadMagYCmd =
{
	.name        = "mpu-read-mag-y",
	.description = "Read magnetometer y value ",
	.usage       = "usage: %s",
	.function    = mpuReadMagYFn
};


static int mpuReadMagZFn (const Command* command, uint argc, char* argv[])
{
	return mpuReadMagAxis (2, command, argc, argv);
}
static Command mpuReadMagZCmd =
{
	.name        = "mpu-read-mag-z",
	.description = "Read magnetometer z value",
	.usage       = "usage: %s",
	.function    = mpuReadMagZFn
};


static int mpuMagId (void)
{
	int result = mpuMagOpenInit();
	if (result < 0)
		return result;

	result = pixi_mpuMagReadRegister (0);
	if (result < 0)
		return result;

	printf ("%02x\n", result);

	pixi_mpuClose();

	return result;
}

static int mpuMagIdFn (const Command* command, uint argc, char* argv[])
{
	LIBPIXI_UNUSED(argv);
	if (argc != 1)
		return commandUsageError (command);

	return mpuMagId();
}
static Command mpuMagIdCmd =
{
	.name        = "mpu-read-mag-id",
	.description = "Read MPU magnetometer ID",
	.usage       = "usage: %s",
	.function    = mpuMagIdFn,
	.flags       = CmdHidden
};


static const Command* commands[] =
{
	&mpuGetAccelScaleCmd,
	&mpuSetAccelScaleCmd,
	&mpuGetGyroScaleCmd,
	&mpuSetGyroScaleCmd,
	&mpuMonitorTempCmd,
	&mpuMonitorMotionCmd,
	&mpuMonitorMagCmd,
	&mpuReadTempCmd,
	&mpuReadAccCmd,
	&mpuReadAccXCmd,
	&mpuReadAccYCmd,
	&mpuReadAccZCmd,
	&mpuReadGyroCmd,
	&mpuReadGyroXCmd,
	&mpuReadGyroYCmd,
	&mpuReadGyroZCmd,
	&mpuReadMagCmd,
	&mpuReadMagXCmd,
	&mpuReadMagYCmd,
	&mpuReadMagZCmd,
	&mpuMagIdCmd,
};


static CommandGroup pixiMpuGroup =
{
	.name      = "pixi-mpu",
	.count     = ARRAY_COUNT(commands),
	.commands  = commands,
	.nextGroup = NULL
};

static void PIO_CONSTRUCTOR (1080) initGroup (void)
{
	addCommandGroup (&pixiMpuGroup);
}
